src/mesh/amr_loadbalance.cpp

```diff
@@ -45,6 +45,241 @@
+  const int ox2 = ((fine_loc.lx2 & 1LL) == 1LL);
+  const int ox3 = ((fine_loc.lx3 & 1LL) == 1LL);
+  int tag = CreateAMRMPITag(lid_recv, ox1, ox2, ox3);
+  if (var->IsAllocated()) {
+    PARTHENON_MPI_CHECK(MPI_Isend(var->coarse_s.data(), var->coarse_s.size(),
+                                  MPI_PARTHENON_REAL, dest_rank, tag, comm, &req));
+  } else {
+    PARTHENON_MPI_CHECK(MPI_Isend(var->coarse_s.data(), 0, MPI_PARTHENON_REAL, dest_rank,
+                                  tag, comm, &req));
+  }
+  return req;
+}
+#endif
+
+bool TryRecvFineToCoarse(int lid_recv, int send_rank, const LogicalLocation &fine_loc,
+                         Variable<Real> *var_in, Variable<Real> *var, MeshBlock *pmb,
+                         Mesh *pmesh) {
+  // TODO(LFR): Set the index ranges appropriately for general topological element
+  // variables
+  static const IndexRange ib = pmb->c_cellbounds.GetBoundsI(IndexDomain::interior);
+  static const IndexRange jb = pmb->c_cellbounds.GetBoundsJ(IndexDomain::interior);
+  static const IndexRange kb = pmb->c_cellbounds.GetBoundsK(IndexDomain::interior);
+
+  const int ox1 = ((fine_loc.lx1 & 1LL) == 1LL);
+  const int ox2 = ((fine_loc.lx2 & 1LL) == 1LL);
+  const int ox3 = ((fine_loc.lx3 & 1LL) == 1LL);
+
+  const int ks = (ox3 == 0) ? 0 : (kb.e - kb.s + 1);
+  const int js = (ox2 == 0) ? 0 : (jb.e - jb.s + 1);
+  const int is = (ox1 == 0) ? 0 : (ib.e - ib.s + 1);
+
+  int test = 1;
+#ifdef MPI_PARALLEL
+  MPI_Comm comm = pmesh->GetMPIComm(var->label());
+  int tag = CreateAMRMPITag(lid_recv, ox1, ox2, ox3);
+  MPI_Status status;
+  PARTHENON_MPI_CHECK(MPI_Iprobe(send_rank, tag, comm, &test, &status));
+#endif
+  if (test) {
+    int size = var_in->IsAllocated();
+#ifdef MPI_PARALLEL
+    PARTHENON_MPI_CHECK(MPI_Get_count(&status, MPI_PARTHENON_REAL, &size));
+#endif
+    if (size > 0) {
+      if (!pmb->IsAllocated(var->label())) pmb->AllocateSparse(var->label());
+      auto cb = var_in->coarse_s;
+#ifdef MPI_PARALLEL
+      // This has to be an MPI_Recv w/o buffering
+      PARTHENON_MPI_CHECK(MPI_Recv(var->coarse_s.data(), var->coarse_s.size(),
+                                   MPI_PARTHENON_REAL, send_rank, tag, comm,
+                                   MPI_STATUS_IGNORE));
+      cb = var->coarse_s;
+#endif
+      auto fb = var->data;
+      const int nt = fb.GetDim(6) - 1;
+      const int nu = fb.GetDim(5) - 1;
+      const int nv = fb.GetDim(4) - 1;
+      parthenon::par_for(
+          DEFAULT_LOOP_PATTERN, "FillSameRankCoarseToFineAMR", DevExecSpace(), 0, nt, 0,
+          nu, 0, nv, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+          KOKKOS_LAMBDA(const int t, const int u, const int v, const int k, const int j,
+                        const int i) {
+            fb(t, u, v, k + ks, j + js, i + is) = cb(t, u, v, k, j, i);
+          });
+      // We have to block here w/o buffering so that the write is guaranteed to be
+      // finished before we get here again
```

Comment:
Why/how could we get in here again?

Reply:
When writing from fine to coarse, there are eight old fine blocks that write to a single coarse block. Each of these correspond to a `TryRecvFineToCoarse` call that does an `MPI_Recv` into the same coarse buffer. If the set kernel hasn't finished, the coarse buffer data can be overwritten before it is copied over. I have updated the comment to be more clear about this.

This implementation was meant to minimize memory usage, but this may not be the ideal from a performance perspective. If this becomes an issue, we could make a pool of buffers that we can pull from and only release them at the end of `RedistributeAndRefineMeshBlocks` (after a `Kokkos::fence()`).

Reply:
Understood. Thanks for the clarification.